Question: Is this implementation of foldLocal valid?

dsyme opened the pull request:
Hi @paladin, @eiriktsarpalis 

I'm looking at the CloudStream implementation out of curiosity.  Is this kind of implementation of "foldLocal" that allows the map and collect functions to return ``Local<T>`` valid?

It uses

```
            let! ctx = Cloud.GetExecutionContext()
            let run a = Async.RunSync(Cloud.ToAsync(a, ctx.Resources,ctx.CancellationToken))
```

to run the local fragments, is there a better way to do this?

Thanks
Don

eiriktsarpalis commented:
You could use `Cloud.RunSynchronously` directly. This is also found in the `MBrace.Continuations` namespace.

palladin commented:
You can check this 
https://github.com/mbraceproject/MBrace.Core/blob/master/src/MBrace.Streams/CloudStreams.fs#L141 for an analogous technique.

dsyme commented:
OK, thanks, I updated this with possible implementations for mapLocal etc.  Please take a look and let me know if you think this is a good thing to be doing. No tests yet - they would need to be added, so this is WIP.

palladin commented:
Combinators like mapLocal, filterLocal, foldLocal etc are certainly useful for many practical scenarios like Cloud.Log in the middle of a pipeline and in general monadic operations in the middle of a pipeline.

dsyme commented:
I updated this with the complete set of Local operators for the file.

I also added sortByDescending and sortByUsing (takes an IComparer) and fixed the sorting semantics so that it uses F# sorting (LanguagePrimitives.FastGenericComparer)

palladin commented:
Great work Don! It is great to have a complete set of Local operators! I was planning to implement all the xxxLocal combinators... so thanks for the contribution!

palladin commented:
@dsyme is it ok with you to merge this pull request?

dsyme commented:
I'd like to add testing for it (if necessary I'll add some test scaffolding)

palladin commented:
Ok cool, I'll wait for your tests and let me know when you are done.

eiriktsarpalis commented:
I'd like to push a nuget update, should I wait for the PR to be merged first?

palladin commented:
I think it is ok to release a nuget package. The major new addition for this nuget release is the ofTextFileByLine combinator.

palladin commented:
@dsyme is it ok with you to merge this pull request? (and I'll add all the missing unit-tests)

dsyme commented:
Yes, that's fine, I'm confident in the PR, thanks for doing the unit tests!

palladin commented:
Thanks Don!
